cardinal/config.go

```diff
@@ -43,47 +51,68 @@ var (
 		TelemetryTraceAddress:     "",
 	}
+
+	// once is used for defining flags once
+	once sync.Once
 )
 
 type WorldConfig struct {
 	// CardinalNamespace The shard namespace for Cardinal. This needs to be unique to prevent signature replay attacks.
-	CardinalNamespace string `config:"CARDINAL_NAMESPACE"`
+	CardinalNamespace string `mapstructure:"CARDINAL_NAMESPACE"`
 
 	// CardinalRollupEnabled When true, Cardinal will sequence and recover to/from base shard.
-	CardinalRollupEnabled bool `config:"CARDINAL_ROLLUP_ENABLED"`
+	CardinalRollupEnabled bool `mapstructure:"CARDINAL_ROLLUP_ENABLED"`
 
 	// CardinalLogLevel Determines the log level for Cardinal.
-	CardinalLogLevel string `config:"CARDINAL_LOG_LEVEL"`
+	CardinalLogLevel string `mapstructure:"CARDINAL_LOG_LEVEL"`
 
 	// CardinalLogPretty Pretty logging, disable by default due to performance impact.
-	CardinalLogPretty bool `config:"CARDINAL_LOG_PRETTY"`
+	CardinalLogPretty bool `mapstructure:"CARDINAL_LOG_PRETTY"`
 
 	// RedisAddress The address of the redis server, supports unix sockets.
-	RedisAddress string `config:"REDIS_ADDRESS"`
+	RedisAddress string `mapstructure:"REDIS_ADDRESS"`
 
 	// RedisPassword The password for the redis server. Make sure to use a password in production.
-	RedisPassword string `config:"REDIS_PASSWORD"`
+	RedisPassword string `mapstructure:"REDIS_PASSWORD"`
 
 	// BaseShardSequencerAddress This is the address that Cardinal will use to sequence and recover to/from base shard.
-	BaseShardSequencerAddress string `config:"BASE_SHARD_SEQUENCER_ADDRESS"`
+	BaseShardSequencerAddress string `mapstructure:"BASE_SHARD_SEQUENCER_ADDRESS"`
 
 	// BaseShardRouterKey is a token used to secure communications between the game shard and the base shard.
-	BaseShardRouterKey string `config:"BASE_SHARD_ROUTER_KEY"`
+	BaseShardRouterKey string `mapstructure:"BASE_SHARD_ROUTER_KEY"`
 
 	// TelemetryEnabled When true, Cardinal will send telemetry to a telemetry agent.
-	TelemetryEnabled bool `config:"TELEMETRY_ENABLED"`
+	TelemetryEnabled bool `mapstructure:"TELEMETRY_ENABLED"`
 
 	// TelemetryStatsdAddress The address of a statsd metric agent that will collect stats from Cardinal.
-	TelemetryStatsdAddress string `config:"TELEMETRY_STATSD_ADDRESS"`
+	TelemetryStatsdAddress string `mapstructure:"TELEMETRY_STATSD_ADDRESS"`
 
 	// TelemetryTraceAddress The address of an agent that supports the collection of traces (e.g. a DataDog agent).
-	TelemetryTraceAddress string `config:"TELEMETRY_TRACE_ADDRESS"`
+	TelemetryTraceAddress string `mapstructure:"TELEMETRY_TRACE_ADDRESS"`
 }
 
 func loadWorldConfig() (*WorldConfig, error) {
+	// Set default config
 	cfg := defaultConfig
 
-	if err := config.FromEnv().To(&cfg); err != nil {
+	// Setup Viper for world toml config file
+	setupViper()
+
+	// Read the config file
+	if err := viper.ReadInConfig(); err != nil {
+		log.Warn().Err(err).Msg("Failed to read config file")
+		log.Info().Msg("Using default config values")
```

Comment:
you can just collapse this message into the warn log

Reply:
updated